src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -775,18 +763,57 @@ private static void finish(@CheckForNull final String cookie) {
         private static final class Callback extends BodyExecutionCallback.TailCall {
 
             private final String cookie;
+            @Deprecated
             private WorkspaceList.Lease lease;
+            private final ExecutorStepExecution execution;
 
-            Callback(String cookie, WorkspaceList.Lease lease) {
+            Callback(String cookie, ExecutorStepExecution execution) {
                 this.cookie = cookie;
-                this.lease = lease;
+                this.execution = execution;
             }
 
             @Override protected void finished(StepContext context) throws Exception {
                 LOGGER.log(FINE, "finished {0}", cookie);
-                lease.release();
-                lease = null;
-                finish(cookie);
+                try {
+                    if (execution != null) {
+                        WorkspaceList.Lease _lease = ExtensionList.lookupSingleton(ExecutorStepDynamicContext.WorkspaceListLeaseTranslator.class).get(execution.state);
+                        if (_lease != null) {
+                            _lease.release();
+                        }
+                    } else {
+                        lease.release();
+                        lease = null;
+                    }
+                } finally {
+                    finish(cookie);
+                }
+                if (execution != null) {
+                    boolean _stopping = execution.state.task.stopping;
+                    execution.state.task.stopping = true;
+                    try {
+                        Queue.getInstance().cancel(execution.state.task);
+                    } finally {
+                        execution.state.task.stopping = _stopping;
+                    }
+                    execution.state = null;
+                }
+            }
+
+            @Override public void onFailure(StepContext context, Throwable t) {
+                try {
+                    if (execution != null) {
+                        TaskListener listener = context.get(TaskListener.class);
+                        if (ExtensionList.lookup(ExecutorStepRetryEligibility.class).stream().anyMatch(e -> e.shouldRetry(t, execution.state.node, execution.step.getLabel(), listener))) {
```

Comment:
Any thoughts on whether this behavior should be something users are able to disable at a per-step level (e.g. `node(..., allowRetry: false)`)? Or that they have to explicitly enable so that the current behavior does not change unless they want it to?

For example, maybe I have two `node` steps in a build, one of them runs tests and so I am happy to let it retry over and over, but the other performs a release and I would prefer to not retry that one even if some implementation of `ExecutorStepRetryEligibility.shouldRetry` returns `true`.

Reply:
Yeah I have not really decided here. Making it opt-in would degrade the value quite a bit because you could not expect to just restore a whole cluster from backup (or whatever) and expect builds to resume without Pipeline modifications. Making it opt-out on a per-`node` basis could make sense.

Reply:
Yeah, maybe we could add a system property that works as a global opt-out in case someone runs into issues after updating.

Reply:
Something like that. The possibilities both for opt-in and opt-out include

* a system property
* some `GlobalConfiguration` setting
* a parameter to the `node` step (also e.g. `KubernetesDeclarativeAgent`)

And not just _whether_ to retry, but if so, some maximum number of retries (so 0 to disable).

Safest would of course be to make all of this stuff be disabled by default (opt-in) initially. However in order for retry to work across restarts, I needed to suppress the various `Pickle`s in favor of `ExecutorStepDynamicContext`, which is a rather intrusive change (most of this PR); and conditionally using `ExecutorStepDynamicContext` but otherwise using `Pickle`s seemed like a nightmare to understand and test. I am fairly well convinced that moving away from `Pickle`s (at least those that need `TryRepeatedly`) and passing control over resumption to `StepExecution.onResume` implementations is the right thing to do in general—this goes back to a debate I had with @kohsuke in 2014—but it is pretty high risk. I do not expect to open this up for merge until January.